### PR TITLE
Update RWDT and refactor RTC

### PR DESCRIPTION
**Pull request by JurajSadel**
- update and add RTC example
- refactor RTC and add a very modified example

Regarding RTC:
This code should give similar calibration values (`cal_val`) as ESP-IDF does except for `RtcCal32kRc`.
I'm really not sure if we want to add/keep the `clock_monitor` example and also the `estimate_xtal_frequency()` function - I let it here just fur testing purposes, besides that, I don't think it's useful tbh and it also crashes with `external` clock sources because they return `cal_val = 0`

Probably the whole RTC will need some additional cleanup and fixes but we can address it after we get the initial PR into the main.

We can discuss how to proceed further with RTC here.

Note: For some reason, I have troubles patching the `LP_WDT` interrupt even if it works for Jesse fine locally - I'll look into that later so the CI will not be happy with `rtc_watchdog` example

### CODE REVIEW

File: esp-hal-common/src/rtc_cntl/mod.rs [hunk 1]
```diff
@@ -106,6 +99,7 @@ impl Clock for RtcSlowClock {
 }
 
 #[cfg(not(esp32c6))]
+#[allow(unused)]
```
Comment (jessebraham) [inline esp-hal-common/src/rtc_cntl/mod.rs]: Why do we need `#[allow(unused)]` here? If this type is not used by a certain chip then it should be `cfg`'d away for that chip.

File: esp-hal-common/src/rtc_cntl/mod.rs [hunk 2]
```diff
@@ -121,6 +115,7 @@ pub(crate) enum RtcCalSel {
 }
 
 #[cfg(esp32c6)]
+#[allow(dead_code)]
```
Comment (jessebraham) [inline esp-hal-common/src/rtc_cntl/mod.rs]: Same as above.

File: esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs [hunk 3]
```diff
@@ -104,165 +86,59 @@ pub(crate) fn configure_clock() {
...
-// }
+fn modem_clk_domain_active_state_icg_map_preinit() {
+    unsafe {
+        let pmu = &*PMU::PTR;
+        let lp_clkrst = &*LP_CLKRST::PTR;
+
+        pmu.hp_active_icg_modem.modify(|_, w| w.hp_active_dig_icg_modem_code().bits(2));
+
+        const MODEM_SYSCON_CLK_CONF_POWER_ST: u32 = 0x600A9800 + 0xc;
+        const MODEM_LPCON_CLK_CONF_POWER_ST: u32 = 0x600A9800 + 0x20;
```
Comment (jessebraham) [inline esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs:114]: I don't seem to even have register description files for these peripherals, so we'll need to update this at a later date. This is fine as is, just wanted to mention that.

File: esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs [hunk 4]
```diff
@@ -315,3 +191,498 @@ pub enum SocResetReason {
     /// JTAG resets CPU
     Cpu0JtagCpu   = 0x18,
 }
+
+extern "C" {
+    fn ets_delay_us(us: u32);
+    fn rtc_get_reset_reason(cpu_num: u32) -> u32;
+}
+
+#[allow(unused)]
```
Comment (jessebraham) [inline esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs:204]: Again, why does this type exist if it's never used?

**Review by jessebraham — COMMENTED** [review]
Seems fine in general, though there's a lot of low-level code here so only so much review I can do.

I'm fine with just ignoring the `clock_monitor` example for now, for the sake of wrapping this up.

@mabezdev @bjoernq would you mind taking a peek at this if you get a chance?

**Comment by jessebraham** [comment]
I've updated `esp32c6` to include the `LP_WDT` interrupt with https://github.com/esp-rs/esp-pacs/pull/96

**Comment by bjoernQ** [comment]
I think there are some `#[allow(unused)]` etc. left over from before moving the code to `esp32c6.rs` , so probably worth to double check that

besides that, I think I followed this topic a bit and it seems this is really the best we can come up with for now

so just those minor things and we should be ready for the next step with ESP32-C6 - even if not everything is perfect yet it will probably help to get it into `main`

File: esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs [hunk 5]
```diff
@@ -315,3 +196,500 @@ pub enum SocResetReason {
...
+
+        let dig_32k_xtal_enabled = lp_clkrst.clk_to_hp.read().icg_hp_xtal32k().bit_is_set();
+
+        if cal_clk == RtcCalSel::RtcCal32kXtal && !dig_32k_xtal_enabled {
+            lp_clkrst
+                .clk_to_hp
+                .modify(|_, w| w.icg_hp_xtal32k().set_bit());
+        }
+
+        // TODO: very hacky
```
Comment (MabezDev) [inline esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs:447]: What is this comment referring to? Can we get an explanation as to why the following is hacky and what needs to be done to improve it in the future?

File: esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs [hunk 6]
```diff
@@ -315,3 +196,500 @@ pub enum SocResetReason {
...
+
+            if !dig_rc32k_enabled {
+                lp_clkrst
+                    .clk_to_hp
+                    .modify(|_, w| w.icg_hp_osc32k().set_bit());
+            }
+        }
+
+        // Check if there is already running calibration process
+        let timg0 = unsafe { &*TIMG0::ptr() };
```
Comment (MabezDev) [inline esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs:507]: We should probably put a TODO here to &mut TIMG0 for calibration?

File: esp32c6-hal/examples/clock_monitor.rs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    let system = peripherals.PCR.split();
+    let clocks = ClockControl::boot_defaults(system.clock_control).freeze();
+
+    let mut rtc = Rtc::new(peripherals.LP_CLKRST);
+
+    // Disable watchdog timers
+    rtc.swd.disable();
+    rtc.rwdt.disable();
+
+    //rtc.rwdt.start(2000u64.millis());
```
Comment (MabezDev) [inline esp32c6-hal/examples/clock_monitor.rs]: We should remove these unused comments

**Review by MabezDev — APPROVED** [review]
Overall LGTM, just a couple of points to address from me :)

**Comment by MabezDev** [comment]
It seems me and @bjoernQ figured out the direct boot issue at the same time :D - I opened https://github.com/jessebraham/esp-hal/pull/4 which includes the same fix as Bjoern's here, plus a couple of binary size improvements.

**Comment by bjoernQ** [comment]
> It seems me and @bjoernQ figured out the direct boot issue at the same time :D - I opened #4 which includes the same fix as Bjoern's here, plus a couple of binary size improvements.

haha what a coincidence - seems like my fix is not enough to also make pcnt_encoder work in direct-boot mode
we should first get yours in and then look into pcnt_encoder (if that is still an issue after that PR)

**Comment by bjoernQ** [comment]
Gave the changes from #4 a try and with that all examples build in direct-boot mode - however pcnt_encoder doesn't work yet

**Comment by MabezDev** [comment]
The error I get for pcnt_encoder is a weird one, I get `already borrowed: BorrowMutError`, which I guess means `critical_section` impl isn't working correctly?

**Comment by bjoernQ** [comment]
> The error I get for pcnt_encoder is a weird one, I get `already borrowed: BorrowMutError`, which I guess means `critical_section` impl isn't working correctly?

I _think_ that some garbage gets copied into `.data` and that makes it believe it's already borrowed (or it's the right data copied to a wrong place)

**Comment by jessebraham** [comment]
I've merged #4, @JurajSadel could you please rebase?

**Comment by MabezDev** [comment]
I think I fixed it in #4, mind trying again?

**Comment by bjoernQ** [comment]
> I think I fixed it in #4, mind trying again?

Seems like that helped! (I still haven't found my rotary encoder but I don't see the crash anymore) 🎉

**Comment by JurajSadel** [comment]
I _think_ this should be good to go, [CI is green](https://github.com/JurajSadel/esp-hal/actions/runs/4251214060)

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -138,14 +138,17 @@ jobs:
         run: cd esp32c6-hal/ && cargo build --examples
       - name: build esp32c6-hal (direct-boot)
         run: cd esp32c6-hal/ && cargo build --examples --features=direct-boot
+      # FIXME: Building using the mcu-boot format currently results in an error.
+      # - name: build esp32c6-hal (mcu-boot)
+      #   run: cd esp32c6-hal/ && cargo build --examples --features=mcu-boot
```
Comment (jessebraham) [inline .github/workflows/ci.yml]: We don't support this image format for this chip, so this can be removed.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -138,14 +138,17 @@ jobs:
...
       # confident that they link.
       - name: check esp32c6-hal (common features)
-        run: cd esp32c6-hal/ && cargo check --examples --features=eh1,ufmt
+        run: cargo check --manifest-path=esp32c6-hal/Cargo.toml --target=riscv32imac-unknown-none-elf --examples --features=eh1,smartled,ufmt
       - name: check esp32c6-hal (async, systick)
-        run: cd esp32c6-hal/ && cargo check --example=embassy_hello_world --features=embassy,embassy-time-systick
+        run: cargo check --manifest-path=esp32c6-hal/Cargo.toml --target=riscv32imac-unknown-none-elf --example=embassy_hello_world --features=embassy,embassy-time-systick
       - name: check esp32c6-hal (async, timg0)
-        run: cd esp32c6-hal/ && cargo check --example=embassy_hello_world --features=embassy,embassy-time-timg0
+        run: cargo check --manifest-path=esp32c6-hal/Cargo.toml --target=riscv32imac-unknown-none-elf --example=embassy_hello_world --features=embassy,embassy-time-timg0
```
Comment (jessebraham) [inline .github/workflows/ci.yml]: Please change these back to match the other chips' configurations.

File: esp32c6-hal/Cargo.toml [hunk 10]
```diff
@@ -75,12 +75,15 @@ required-features = ["eh1"]
...
-required-features = ["async", "embassy"]
+required-features = ["embassy"]
 
 [[example]]
 name              = "embassy_wait"
-required-features = ["async", "embassy"]
+required-features = ["embassy", "async"]
+
+# [profile.dev]
+# opt-level = 1
```
Comment (jessebraham) [inline esp32c6-hal/Cargo.toml]: Please remove this.

**Review by jessebraham — COMMENTED** [review]
Just a few things to clean up, looks good otherwise. I'll go test some examples in the meantime.

File: esp32c6-hal/Cargo.toml (hunk 10 above)
Comment (MabezDev) [reply esp32c6-hal/Cargo.toml]: Can we also remove it from the C3 Cargo.toml, seems it was accidentally committed https://github.com/esp-rs/esp-hal/blame/main/esp32c3-hal/Cargo.toml#L86.

File: esp-hal-common/src/clock/clocks_ll/esp32c6.rs [hunk 11]
```diff
@@ -41,7 +41,7 @@ const MODEM_LPCON_CLK_I2C_MST_FO: u32 = 1 << 2;
 const MODEM_LPCON_I2C_MST_CLK_CONF_REG: u32 = DR_REG_MODEM_LPCON_BASE + 0x10;
 const MODEM_LPCON_CLK_I2C_MST_SEL_160M: u32 = 1 << 0;
 
-pub(crate) fn esp32c6_rtc_bbpll_configure(xtal_freq: XtalClock, pll_freq: PllClock) {
+pub(crate) fn esp32c6_rtc_bbpll_configure(_xtal_freq: XtalClock, _pll_freq: PllClock) {
     unsafe {
         // enable i2c mst clk by force on temporarily
         (MODEM_LPCON_CLK_CONF_FORCE_ON_REG as *mut u32).write_volatile(
```
Comment (MabezDev) [inline esp-hal-common/src/clock/clocks_ll/esp32c6.rs:47]: Are these not available in the PAC?

**Comment by jessebraham** [comment]
@JurajSadel the examples I tested (including `rtc_watchdog`) seem to be working, so looks good there. In addition to my comments, could you please also address the earlier comments made by @MabezDev?

**Review by MabezDev — CHANGES_REQUESTED** [review]

File: esp-hal-common/src/clock/clocks_ll/esp32c6.rs (hunk 11 above)
Comment (jessebraham) [reply esp-hal-common/src/clock/clocks_ll/esp32c6.rs:47]: We're still missing register description files for a number of peripherals.

File: esp32c6-hal/Cargo.toml (hunk 10 above)
Comment (jessebraham) [reply esp32c6-hal/Cargo.toml]: Sure, let's do that in a different PR, though.

**Review by MabezDev — APPROVED** [review]

**Review by jessebraham — APPROVED** [review]
CI is green and the examples are working as expected, so I think we're good to go here! Thank you again @JurajSadel for all your work on this, I know it was challenging!

**Comment by JurajSadel** [comment]
@bjoernQ @jessebraham @MabezDev Thank you guys for all the help, reviews and testing!

**Pull request merged**